### PR TITLE
Ao3 5022 rails 4 dot 1

### PR DESCRIPTION
**Pull request by WendyBeth**
# Pull Request Checklist

* [x] Have you read [How to write the perfect pull request](https://github.com/blog/1943-how-to-write-the-perfect-pull-request)?
* [x] Have you read through the [contributor guidelines](https://github.com/otwcode/otwarchive/blob/master/CONTRIBUTING.md)?
* [x] Have you added tests for any changed functionality?
* [x] Have you added the JIRA issue number as the *first* thing in your pull request title (eg: `AO3-1234 Fix thing`)
* [x] Have you updated the JIRA issue with the information below?

## Issue

https://otwarchive.atlassian.net/browse/AO3-5022

## Purpose

What does this PR do?

This PR upgrades the AO3 application to run on Rails 4.1.16. This encompasses many changes including replacing outdated AR query syntax.

## Testing

How can the Archive's QA team verify that this is working as you intended? (If you have access, please copy this into the JIRA ticket for them!)

Due to the wide scope of these changes, a full run-through of site features is required to validate this upgrade.

## References

Are there any other relevant issues / PRs / mailing lists discussions related to this?

No explicit external references.

## Credit

*What name do you want us to use to credit your work in the [Archive of Our Own's Release Notes](http://archiveofourown.org/admin_posts?tag=1)?*

Wendy Randquist - Littlelines

*What pronouns do you prefer (she/her, he/him, zie/hir etc)?*

she/her

(if you do not fill this in, we will use your GitHub account name and will use "they" to avoid making assumptions about your gender)

### CODE REVIEW

File: app/views/admin/admin_users/confirm_delete_user_creations.html.erb [hunk 1]
```diff
@@ -8,20 +8,20 @@
   <p class="caution notice">
     <%= ts("Are you sure you want to <strong><em>delete</em></strong> all the works and comments created by this user,
            along with their %{bookmarks} bookmarks, %{series} series, and %{collections} collections? This <strong>cannot be undone</strong>.",
-           bookmarks: @bookmarks.count, series: @series.count, collections: @collections.count).html_safe %>
+           bookmarks: @bookmarks.size, series: @series.size, collections: @collections.size).html_safe %>
   </p>
 
-  <% if @works.count > 0 %>
```
Comment (zz9pzza) [inline app/views/admin/admin_users/confirm_delete_user_creations.html.erb]: Should these be unless @works.size.zero?

( This really is a style question, I don't know what the accepted practice is )

File: app/controllers/external_authors_controller.rb [hunk 2]
```diff
@@ -76,7 +76,7 @@ def update
     end
     @invitation.mark_as_redeemed if @invitation && !params[:imported_stories].blank?
 
-    if @external_author.update_attributes(external_author_params[:external_author])
+    if @external_author.update_attributes(external_author_params[:external_author] || {})
```
Comment (zz9pzza) [inline app/controllers/external_authors_controller.rb]: Why is the "|| {}" here ?

Comment (WendyBeth) [inline app/controllers/external_authors_controller.rb]: Rails previously returned an empty hash for nested params that were empty - now it returns nil. `update_attributes` will raise an error if it doesn't receive a hash.

File: app/views/admin/admin_users/confirm_delete_user_creations.html.erb (hunk 1 above)
Comment (WendyBeth) [inline app/views/admin/admin_users/confirm_delete_user_creations.html.erb]: I think `@works.size.zero?` would be better, yeah, so I'll go ahead and change it.

File: app/controllers/external_authors_controller.rb (hunk 2 above)
Comment (zz9pzza) [inline app/controllers/external_authors_controller.rb]: Sorry, I understand ( I missed the closing bracket so I thought it was associated with the if)

**Pull request merged**